cpp/src/reader/table_query_executor.cc

```
@@ -85,6 +87,115 @@ int TableQueryExecutor::query(const std::string &table_name,
     return ret;
 }
 
-void TableQueryExecutor::destroy_query_data_set(ResultSet *qds) { delete qds; }
+int TableQueryExecutor::query_on_tree(
+    const std::vector<std::shared_ptr<IDeviceID>>& devices,
+    const std::vector<std::string>& tag_columns,
+    const std::vector<std::string>& field_columns, Filter* time_filter,
+    ResultSet*& ret_qds) {
+    common::PageArena pa;
+    pa.init(512, common::MOD_TSFILE_READER);
+    int ret = common::E_OK;
+    TsFileMeta* file_meta = tsfile_io_reader_->get_tsfile_meta();
+    std::vector<MetaIndexNode*> table_inodes;
+    for (auto const& device : devices) {
+        MetaIndexNode* table_inode;
+        if (RET_FAIL(file_meta->get_table_metaindex_node(
+                device->get_table_name(), table_inode))) {
+        };
+        table_inodes.push_back(table_inode);
+    }
+
+    std::vector<common::ColumnSchema> col_schema;
+    for (auto const& tag : tag_columns) {
+        col_schema.emplace_back(tag, common::TSDataType::STRING,
+                                common::ColumnCategory::TAG);
+    }
+
+    std::unordered_map<std::string, common::TSDataType> column_types_map;
+
+    for (auto const& device : devices) {
+        bool all_collected = true;
+        for (const auto& field_col : field_columns) {
+            if (column_types_map.find(field_col) == column_types_map.end()) {
+                all_collected = false;
+                break;
+            }
+        }
+        if (all_collected) {
+            break;
+        }
+
+        std::unordered_set<std::string> measurements(field_columns.begin(),
+                                                     field_columns.end());
+        std::vector<ITimeseriesIndex*> index(measurements.size());
+        if (RET_FAIL(tsfile_io_reader_->get_timeseries_indexes(
+                device, measurements, index, pa))) {
+            assert(0);
+        }
```

Should return the failure?